### PR TITLE
Changes for production

### PR DESCRIPTION
**Pull request by spacemansteve**
Mostly changes to meta tags, also support to process a file of bibcodes.

### CODE REVIEW

File: adstb/app.py [hunk 1]
```diff
@@ -116,13 +116,18 @@ def _update_timestamps(self, message):
 
     def _load_url(self, url):
         try:
+            self.logger.info('trying to load url {} at endpoint: {}'.format(url, self.conf.get('PUPPETEER_ENDPOINT')))
```
Comment (romanchyla) [inline adstb/app.py]: please turn them to `logger.debug` calls

File: adstb/app.py [hunk 2]
```diff
@@ -294,45 +301,38 @@ def _retrieve_abstract_template(self, url):
...
-        end = html.find('data-highwire', x)
-        while html.find('data-highwire', end+1) > 0:
-            end = html.find('data-highwire', end+1)
-
-        
-        while html[end] != '>':
-            end += 1
-        
-            
+        end = html.find('</head')
```
Comment (romanchyla) [inline adstb/app.py]: this is likely unsafe; you may be removing too much stuff (if you remove something important, bbb will be broken)

File: run.py [hunk 3]
```diff
@@ -85,6 +83,8 @@ def submit(bibcode):
             app.logger.info('Done: {0}/{1}'.format(start, j['response']['numFound']))
             print 'Done submitting: {0}/{1}'.format(start, j['response']['numFound'])
         except Exception, e:
+            import pdb
```
Comment (romanchyla) [inline run.py]: to be removed

File: run.py [hunk 4]
```diff
@@ -119,18 +117,22 @@ def harvest_by_null(queue='priority-bumblebee',
...
             last_id = -1 
-                
-    
+    app.logger.info('harvest_by_null running with last_id of {}'.format(last_id))
+
     while True:
-        r = client.get(url, params=params)
+        # reuse the http ads api client with keep-alive connections and bearer token
+        r = app._client.get(url, params=params)
+        import pdb
```
Comment (romanchyla) [inline run.py]: dtto

File: run.py [hunk 5]
```diff
@@ -119,18 +117,22 @@ def harvest_by_null(queue='priority-bumblebee',
...
-            msg = TurboBeeMsg(target=d['target'],
-                              qid=d['qid'])
-            
+            if iscachable(d['target']):
+                msg = TurboBeeMsg(target=d['target'],
+                                  qid=d['qid'])
             
-            tasks.task_priority_queue.delay(msg)
+                app.logger.info('harvest_by_null queuing target {}'.format(d['target']))
+                tasks.task_priority_queue.delay(msg)
```
Comment (romanchyla) [inline run.py:130]: here i'd recommend using logging to warn the user something, that was explicitly requested, was rejected and won't be processed

File: run.py [hunk 6]
```diff
@@ -158,6 +160,17 @@ def harvest_by_null(queue='priority-bumblebee',
     app.logger.info('Done submitting {0} pages.'.format(i))
     print i, last_id
 
+    
+def iscachable(target):
```
Comment (romanchyla) [inline run.py]: if you make this part of the application (method), then various calls can provide/override this filter - just a thought

**Review by romanchyla — COMMENTED** [review]
please check the unittests and see some comments below, at least one of them - the bbb template `</head>` is important

**Comment by romanchyla** [comment]
please fix the failing unittests @spacemansteve

File: adstb/app.py (hunk 2 above)
Comment (spacemansteve) [reply adstb/app.py]: I don't think I added this, perhaps we should talk about what you'd like instead.

Comment (romanchyla) [reply adstb/app.py]: i think you did, i'm looking at your changes ;-)

my comment is about line 318; if we search for data-highwire, we get the beginning of the tags section -- but there can be lots of stuff between that and the end of the `</head>`

so you are potentially replacing too much; was there a problem with the original lines 317-322?

File: adstb/app.py [hunk 7]
```diff
@@ -341,13 +334,26 @@ def _retrieve_abstract_template(self, url):
...
         if x > end:
-            raise Exception("Cannot find abstract section")
-        
+            raise Exception('Cannot process fetched page, cannot find abstract section for {}'.format(url))
+
         html = html[0:x] + '{{abstract}}' + html[end:]
-        
-        if 'bibcode' in parts and parts['bibcode']:
-            html = html.replace(parts['bibcode'], u'{{bibcode}}')
-
```
Comment (romanchyla) [inline adstb/app.py:350]: this should be kept (and used if it's not already); otherwise links will be wrong

File: adstb/app.py [hunk 8]
```diff
@@ -341,13 +334,26 @@ def _retrieve_abstract_template(self, url):
...
+            raise Exception('Cannot process fetched page, cannot find abstract section for {}'.format(url))
+
         html = html[0:x] + '{{abstract}}' + html[end:]
-        
-        if 'bibcode' in parts and parts['bibcode']:
-            html = html.replace(parts['bibcode'], u'{{bibcode}}')
-        
+
+        # finally, cut out noscript warning that says javascript is required
+        x = html.find('<div id="noscriptmsg"')  # div id within noscript
```
Comment (romanchyla) [inline adstb/app.py]: you may want to search for `id="noscriptmsg"` - that's guaranteed to be formatted that way

File: adstb/bumblebee.py [hunk 9]
```diff
@@ -212,8 +231,39 @@ def build_meta_tags(solrdoc):
...
+    if 'page' in data:
+        data['page'] = data['page'][0]
+    if 'page_range' in data and '-' in data['page_range']:
+        # expected format: 41-59
+        data['lastpage'] = data['page_range'].split('-')[1]
+    if 'title' in data:
+        data['title'] = _html_to_text(data['title'][0])
+    link_url_template = 'http://dev.adsabs.harvard.edu/abs/{}'
+    if 'LINK_URL_TEMPLATE' in conf:
+        data['url'] = conf['LINK_URL_TEMPLATE'].format(data['bibcode'])
```
Comment (romanchyla) [inline adstb/bumblebee.py]: i'm confused by this variable, it is not used in the tags sections; and also, i think it shouldn't be, the template has been generated maybe from a different domain that what's in the config

(which brings up the sensitive point of "should we verify/sanitize links in the template" ?)

**Review by romanchyla — CHANGES_REQUESTED** [review]
please see comments

File: adstb/app.py (hunk 8 above)
Comment (spacemansteve) [reply adstb/app.py]: yep, changed

File: adstb/app.py (hunk 2 above)
Comment (spacemansteve) [reply adstb/app.py]: thx, reverted code back to original

File: adstb/bumblebee.py (hunk 9 above)
Comment (spacemansteve) [reply adstb/bumblebee.py]: updated pr

File: adstb/app.py (hunk 7 above)
Comment (spacemansteve) [reply adstb/app.py:350]: updated pr

File: adstb/bumblebee.py [hunk 10]
```diff
@@ -212,8 +231,34 @@ def build_meta_tags(solrdoc):
...
 
+
+htmlParser = HTMLParser()
+tag_re = re.compile(r'<[^>]+>')
+
+
+def _html_to_text(t):
+    """remove html tags and clean html encoded characters and return plain text string"""
+    t = htmlParser.unescape(t)
+    t = tag_re.sub('', t)
```
Comment (romanchyla) [inline adstb/bumblebee.py:258]: shouldn't the lines 257-258 be swapped?

**Review by romanchyla — APPROVED** [review]
Just a small comment below

Also, can you please check the failing travis? I do realize that the tests are reported as failing for a long time (maybe they never went green?) but it is likely something silly, they should work before me merge.

I'm not seeing any major change in the code; are we satisfied that the current code generates proper static pages for turbobee?

**Comment by romanchyla** [comment]
you sure about this latest change? the template will miss {{bibcode}}
placeholders...

File: adstb/bumblebee.py (hunk 10 above)
Comment (spacemansteve) [reply adstb/bumblebee.py:258]: I like what it currently does, converting u'Renormalization of quark propagators from twisted-mass lattice QCD at N&lt;SUB&gt;f&l\
t;/SUB&gt;=2' to u'Renormalization of quark propagators from twisted-mass lattice QCD at Nf=2'.  What did you have in mind?

**Comment by spacemansteve** [comment]
Checks pass on github after a change to config.py
I'm not sure about "you sure about this latest change? the template will miss {{bibcode}}
placeholders...".  I'll thought you wanted this included.  I will verify after PR has been merged.

**Pull request merged**